cfme/utils/auth/__init__.py

```diff
@@ -0,0 +1,178 @@
+        ep.name: ep.resolve()
+        for ep in iter_entry_points('manageiq.auth_provider_types')
+    }
+
+
+def auth_class_from_type(auth_prov_type):
+    """Using the registered auth provider classes, fetch a class by its type key
+
+    Args:
+        auth_prov_type: string key matching a registered type in entry_points
+
+    Raises:
+        UnknownProviderType when the given type isn't registered in entry_points
+    """
+    try:
+        return auth_provider_types()[auth_prov_type]
+    except KeyError:
+        raise UnknownProviderType('Unknown auth provider type: {}'.format(auth_prov_type))
+
+
+def get_auth_crud(auth_prov_key):
+    """Get a BaseAuthProvider derived class with the auth_data.yaml configuration for the key
+
+    Args:
+        auth_prov_key: string key matching one in conf/auth_data.yaml 'auth_providers' dict
+    Raises:
+        ValueError if the yaml type for given key doesn't match auth_type on fetched class
+    """
+    auth_prov_config = auth_prov_data[auth_prov_key].copy()
+    klass = auth_class_from_type(auth_prov_config.get('type'))
+    if auth_prov_config.get('type') != klass.auth_type:
+        raise ValueError('{} must have type "{}"'.format(klass.__name__, klass.auth_type))
+    return klass.from_config(auth_prov_config, auth_prov_key)
+
+
+@attr.s
+class BaseAuthProvider(object):
+    """Base class for authentication provider objects
+
+    Might not be needed
+    """
+    auth_type = None
+    view_class = None
+    key = attr.ib()
+
+    @classmethod
+    def from_config(cls, prov_config, prov_key):
+        """Returns an object using the passed yaml config
+        Sets defaults for yaml configured objects separate from attr.ib definitions
+        """
+        prov_config.update(credentials[prov_config.get('credentials')])
+        class_attrs = [att.name for att in cls.__attrs_attrs__]
+        class_params = {k: v for k, v in iteritems(prov_config) if k in class_attrs}
+        return cls(key=prov_key, **class_params)
+
+    def as_fill_value(self):
+        class_attrs = [att.name for att in self.__attrs_attrs__]
+        include_attrs = [getattr(self.__class__, name)
+                         for name in self.view_class.cls_widget_names()
+                         if name in class_attrs]
+        return attr.asdict(self, filter=attr.filters.include(*include_attrs))
+
+
+@attr.s
+class AmazonAuthProvider(BaseAuthProvider):
+    """AWS IAM auth provider"""
+    auth_type = 'amazon'
+    view_class = AmazonAuthenticationView
+
+    access_key = attr.ib()
+    secret_key = attr.ib()
+    get_groups = attr.ib(default=False)
+
+
+@attr.s
+class MIQAuthProvider(BaseAuthProvider):
+    """base class for miq auth providers (ldap/ldaps modes in UI)
+    Intended to be used for freeipa, AD, openldap and openldaps type providers
+    """
+    auth_type = None
+    host1 = attr.ib()
+    bind_password = attr.ib()  # Ordered to adhere to mandatory attrs sequence
+    host2 = attr.ib(default=None)
```

Comment:
maybe hosts = [] ?

Reply:
The older auth config actually kind of had a list of hosts - but it created complexities otherwise that made me prefer specifying the hosts as 1/2/3 - this also allows controlling their configuration order a little easier.